reasoner/resolution/resolver/DisjunctionResolver.java

```diff
@@ -0,0 +1,225 @@
+        if (!responseProducers.containsKey(fromUpstream)) {
+            responseProducers.put(fromUpstream, responseProducerCreate(fromUpstream, iteration));
+        } else {
+            ResponseProducer responseProducer = responseProducers.get(fromUpstream);
+            assert responseProducer.iteration() == iteration ||
+                    responseProducer.iteration() + 1 == iteration;
+
+            if (responseProducer.iteration() + 1 == iteration) {
+                // when the same request for the next iteration the first time, re-initialise required state
+                ResponseProducer responseProducerNextIter = responseProducerReiterate(fromUpstream, responseProducer, iteration);
+                responseProducers.put(fromUpstream, responseProducerNextIter);
+            }
+        }
+        return responseProducers.get(fromUpstream);
+    }
+
+    @Override
+    protected ResponseProducer responseProducerCreate(Request fromUpstream, int iteration) {
+        LOG.debug("{}: Creating a new ResponseProducer for request: {}", name(), fromUpstream);
+        assert fromUpstream.partialAnswer().isFiltered() || fromUpstream.partialAnswer().isIdentity();
+        ResponseProducer responseProducer = new ResponseProducer(Iterators.empty(), iteration);
+        assert !downstreamResolvers.isEmpty();
+        for (Actor<ConjunctionResolver.Nested> conjunctionResolver : downstreamResolvers) {
+            Filtered downstream = fromUpstream.partialAnswer()
+                    .filterToDownstream(conjunctionRetrievedIds(conjunctionResolver));
+            Request request = Request.create(self(), conjunctionResolver, downstream);
+            responseProducer.addDownstreamProducer(request);
+        }
+        return responseProducer;
+    }
+
+    @Override
+    protected ResponseProducer responseProducerReiterate(Request fromUpstream, ResponseProducer responseProducerPrevious,
+                                                         int newIteration) {
+        assert newIteration > responseProducerPrevious.iteration();
+        LOG.debug("{}: Updating ResponseProducer for iteration '{}'", name(), newIteration);
+
+        assert newIteration > responseProducerPrevious.iteration();
+
+        ResponseProducer responseProducerNewIter = responseProducerNewIter(responseProducerPrevious, newIteration);
+        for (Actor<ConjunctionResolver.Nested> conjunctionResolver : downstreamResolvers) {
+            Filtered downstream = fromUpstream.partialAnswer()
+                    .filterToDownstream(conjunctionRetrievedIds(conjunctionResolver));
+            Request request = Request.create(self(), conjunctionResolver, downstream);
+            responseProducerNewIter.addDownstreamProducer(request);
+        }
+        return responseProducerNewIter;
+    }
+
+    protected ResponseProducer responseProducerNewIter(ResponseProducer responseProducerPrevious, int newIteration) {
```

Comment:
may as well make this abstract and then implement it separately in `Nested` and `Root` :) overriding implemented methods is a bit meh